core/embed/extmod/modtrezorio/modtrezorio-fatfs.h

```diff
@@ -0,0 +1,524 @@
+  tuple->items[0] = mp_obj_new_int_from_uint(info->fsize);
+  char attrs[] = "-----";
+  if (info->fattrib & AM_RDO) {
+    attrs[0] = 'r';
+  }
+  if (info->fattrib & AM_HID) {
+    attrs[1] = 'h';
+  }
+  if (info->fattrib & AM_SYS) {
+    attrs[2] = 's';
+  }
+  if (info->fattrib & AM_DIR) {
+    attrs[3] = 'd';
+  }
+  if (info->fattrib & AM_ARC) {
+    attrs[4] = 'a';
+  }
+  tuple->items[1] = mp_obj_new_str(attrs, 5);
+  tuple->items[2] = mp_obj_new_str(info->fname, strlen(info->fname));
+  return MP_OBJ_FROM_PTR(tuple);
+}
+
+/// class FatFSFile:
+///     """
+///     Class encapsulating file
+///     """
+typedef struct _mp_obj_FatFSFile_t {
+  mp_obj_base_t base;
+  FIL fp;
+} mp_obj_FatFSFile_t;
+
+/// def __exit__(self) -> None:
+///     """
+///     Close an open file object
+///     """
+STATIC mp_obj_t mod_trezorio_FatFSFile___exit__(size_t n_args,
+                                                const mp_obj_t *args) {
+  mp_obj_FatFSFile_t *o = MP_OBJ_TO_PTR(args[0]);
+  FRESULT res = f_close(&(o->fp));
+  if (res != FR_OK) {
+    mp_raise_OSError(fresult_to_errno_table[res]);
+  }
+  return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(mod_trezorio_FatFSFile___exit___obj,
+                                           4, 4,
+                                           mod_trezorio_FatFSFile___exit__);
+
+/// def close(self) -> None:
+///     """
+///     Close an open file object
+///     """
+STATIC mp_obj_t mod_trezorio_FatFSFile_close(mp_obj_t self) {
+  mp_obj_FatFSFile_t *o = MP_OBJ_TO_PTR(self);
+  FRESULT res = f_close(&(o->fp));
+  if (res != FR_OK) {
+    mp_raise_OSError(fresult_to_errno_table[res]);
+  }
+  return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(mod_trezorio_FatFSFile_close_obj,
+                                 mod_trezorio_FatFSFile_close);
+
+/// def read(self, data: bytearray) -> int:
+///     """
+///     Read data from the file
+///     """
+STATIC mp_obj_t mod_trezorio_FatFSFile_read(mp_obj_t self, mp_obj_t data) {
+  mp_obj_FatFSFile_t *o = MP_OBJ_TO_PTR(self);
+  mp_buffer_info_t buf;
+  mp_get_buffer_raise(data, &buf, MP_BUFFER_WRITE);
+  UINT read;
+  FRESULT res = f_read(&(o->fp), buf.buf, buf.len, &read);
+  if (res != FR_OK) {
+    mp_raise_OSError(fresult_to_errno_table[res]);
+  }
+  return mp_obj_new_int_from_uint(read);
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_2(mod_trezorio_FatFSFile_read_obj,
+                                 mod_trezorio_FatFSFile_read);
+
+/// def write(self, data: bytearray) -> int:
+///     """
+///     Write data to the file
+///     """
+STATIC mp_obj_t mod_trezorio_FatFSFile_write(mp_obj_t self, mp_obj_t data) {
+  mp_obj_FatFSFile_t *o = MP_OBJ_TO_PTR(self);
+  mp_buffer_info_t buf;
+  mp_get_buffer_raise(data, &buf, MP_BUFFER_READ);
+  UINT written;
+  FRESULT res = f_write(&(o->fp), buf.buf, buf.len, &written);
+  if (res != FR_OK) {
+    mp_raise_OSError(fresult_to_errno_table[res]);
+  }
```

Comment:
Maybe we should throw "No space left on device" when not all bytes were written before return for consistency with file operations in python:

      if (written != buf.len) {
        /* no space left on device or free cluster count recorded in FSInfo fell to 0 */
        mp_raise_OSError(MP_ENOSPC);
      }